<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
           "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<meta name="GENERATOR" content="TtH 3.77">
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
 <style type="text/css"> div.p { margin-top: 7pt;}</style>
 <style type="text/css"><!--
 td div.comp { margin-top: -0.6ex; margin-bottom: -1ex;}
 td div.comb { margin-top: -0.6ex; margin-bottom: -.6ex;}
 td div.hrcomp { line-height: 0.9; margin-top: -0.8ex; margin-bottom: -1ex;}
 td div.norm {line-height:normal;}
 span.roman {font-family: serif; font-style: normal; font-weight: normal;} 
 span.overacc2 {position: relative;  left: .8em; top: -1.2ex;}
 span.overacc1 {position: relative;  left: .6em; top: -1.2ex;} --></style>


<title> GimmeMotifs documentation</title>
 
<h1 align="center">GimmeMotifs documentation </h1>

<h3 align="center">Simon van Heeringen </h3>

<h3 align="center">Oct  1, 2010
</h3>

<div class="p"><!----></div>
 
<h1>Contents </h1><a href="#tth_sEc1"
>1&nbsp; Introduction</a><br />
<a href="#tth_sEc2"
>2&nbsp; What's included?</a><br />
<a href="#tth_sEc3"
>3&nbsp; Installation</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.1"
>3.1&nbsp; Installation packages</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.1.1"
>3.1.1&nbsp; Installation on Ubuntu or Debian</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.1.2"
>3.1.2&nbsp; Installation on Fedora</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2"
>3.2&nbsp; Installation from source</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2.1"
>3.2.1&nbsp; Prerequisites</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2.2"
>3.2.2&nbsp; Required packages (Python)</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2.3"
>3.2.3&nbsp; Other required packages</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2.4"
>3.2.4&nbsp; Additional motif prediction programs</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.2.5"
>3.2.5&nbsp; Building from source</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3"
>3.3&nbsp; Configuration</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.1"
>3.3.1&nbsp; Data sources</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.2"
>3.3.2&nbsp; Indexing the genomes</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.3"
>3.3.3&nbsp; Adding gene files</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.4"
>3.3.4&nbsp; The easy way: <tt>add_organism.py</tt></a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.5"
>3.3.5&nbsp; Adding motif prediction tools</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.6"
>3.3.6&nbsp; MotifSampler configuration</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc3.3.7"
>3.3.7&nbsp; Other configuration options</a><br />
<a href="#tth_sEc4"
>4&nbsp; Usage</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.1"
>4.1&nbsp; Quick GimmeMotifs example</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.2"
>4.2&nbsp; GimmeMotifs example</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3"
>4.3&nbsp; Using GimmeMotifs: best practices and tips</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3.1"
>4.3.1&nbsp; GimmeMotifs is multi-threaded</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3.2"
>4.3.2&nbsp; Running time</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3.3"
>4.3.3&nbsp; Intermediate results</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3.4"
>4.3.4&nbsp; Running on FASTA files</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.3.5"
>4.3.5&nbsp; Small input sets</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc4.4"
>4.4&nbsp; Detailed options</a><br />
<a href="#tth_sEc5"
>5&nbsp; Other scripts</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc5.1"
>5.1&nbsp; Input formats</a><br />
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#tth_sEc5.2"
>5.2&nbsp; Descriptions</a><br />
<a href="#tth_sEc6"
>6&nbsp; Acknowledgements</a><br />


<div class="p"><!----></div>
 <h2><a name="tth_sEc1">
1</a>&nbsp;&nbsp;Introduction</h2>
GimmeMotifs is a <i>de novo</i> motif prediction pipeline, especially suited for ChIP-seq datasets. It incorporates several existing motif prediction algorithms in an ensemble method to predict motifs and clusters these motifs using the WIC similarity scoring metric. It is freely available for download and use under the MIT license. If you find GimmeMotifs useful, please cite:

<ul>
<li> van Heeringen SJ and Veenstra GJC, GimmeMotifs: a <i>de novo</i> motif prediction pipeline for ChIP-sequencing experiments, <i>submitted</i>.
<div class="p"><!----></div>
</li>
</ul>
This document describes how to install and use GimmeMotifs, for theoretical details, please see our publication .

<div class="p"><!----></div>
Hopefully this document explains at least the basics of installation and usage, but it's probably far from complete. If you have any further question, please don't hesistate to contact me: <a href="s.vanheeringen@ncmls.ru.nl"><tt>s.vanheeringen@ncmls.ru.nl</tt></a>.

<div class="p"><!----></div>
 <h2><a name="tth_sEc2">
2</a>&nbsp;&nbsp;What's included?</h2>
GimmeMotifs is dependent upon the great work of others: motif prediction algorithms. GimmeMotifs includes several tools by default:

<ul>
<li> BioProspector <a href="#liu_bioprospector:_2001" name="CITEliu_bioprospector:_2001">[1</a>] <a href="http://motif.stanford.edu/distributions/bioprospector/"><tt>http://motif.stanford.edu/distributions/bioprospector/</tt></a>
<div class="p"><!----></div>
</li>

<li> GADEM <a href="#li_gadem:genetic_2009" name="CITEli_gadem:genetic_2009">[2</a>] <a href="http://www.niehs.nih.gov/research/resources/software/gadem/index.cfm"><tt>http://www.niehs.nih.gov/research/resources/software/gadem/index.cfm</tt></a>
<div class="p"><!----></div>
</li>

<li> Improbizer <a href="#ao_environmentally_2004" name="CITEao_environmentally_2004">[3</a>] <a href="http://users.soe.ucsc.edu/~kent/"><tt>http://users.soe.ucsc.edu/~kent/</tt></a>
<div class="p"><!----></div>
</li>

<li> MDmodule <a href="#liu_algorithm_2002" name="CITEliu_algorithm_2002">[4</a>] (included in the MotifRegressor Package) <a href="http://www.math.umass.edu/~conlon/mr.html"><tt>http://www.math.umass.edu/~conlon/mr.html</tt></a>
<div class="p"><!----></div>
</li>

<li> MEME <a href="#bailey_meme_2009" name="CITEbailey_meme_2009">[5</a>] <a href="http://meme.sdsc.edu/"><tt>http://meme.sdsc.edu/</tt></a>
<div class="p"><!----></div>
</li>

<li> MoAn <a href="#valen_discovery_2009" name="CITEvalen_discovery_2009">[6</a>] <a href="http://moan.binf.ku.dk/"><tt>http://moan.binf.ku.dk/</tt></a>
<div class="p"><!----></div>
</li>

<li> MotifSampler <a href="#thijs_higher-order_2001" name="CITEthijs_higher-order_2001">[7</a>,<a href="#thijs_gibbs_2002" name="CITEthijs_gibbs_2002">[8</a>] <a href="http://homes.esat.kuleuven.be/~sistawww/bioi/thijs/download.html"><tt>http://homes.esat.kuleuven.be/~sistawww/bioi/thijs/download.html</tt></a>
<div class="p"><!----></div>
</li>

<li> trawler <a href="#ettwiller_trawler:_2007" name="CITEettwiller_trawler:_2007">[9</a>] <a href="http://ani.embl.de/trawler/"><tt>http://ani.embl.de/trawler/</tt></a>
<div class="p"><!----></div>
</li>
</ul>
All these tools are installed automatically with GimmeMotifs. To use these programs <em>no additional steps have to be taken</em> (except for MotifSampler see <a href="#sec:MotifSampler">3.3.6</a>). <b>Please note: all these programs include their own license and most are free for academic or non-commercial use only! For commercial use of any of these programs, please consult the respective author!</b>

<div class="p"><!----></div>
 <h2><a name="tth_sEc3">
3</a>&nbsp;&nbsp;Installation</h2>
GimmeMotifs runs on Linux. Definitely not on Windows, sorry. Mac OS X should work in theory, but as I don't have the means to test this, I'm not completely sure. I have tried to make installation of GimmeMotifs as easy as possible. There are installation packages available for Fedora, Ubuntu and Debian. It's also possible to install GimmeMotifs from source, but it depends on quite some external packages. Please make sure all prerequisites are installed before installing GimmeMotifs from source. 

<div class="p"><!----></div>
     <h3><a name="tth_sEc3.1">
3.1</a>&nbsp;&nbsp;Installation packages</h3>

      <h4><a name="tth_sEc3.1.1">
3.1.1</a>&nbsp;&nbsp;Installation on Ubuntu or Debian</h4>
Download the latest GimmeMotifs .deb package for your distribution  from <a href="http://www.ncmls.nl/bioinfo/gimmemotifs/"><tt>http://www.ncmls.nl/bioinfo/gimmemotifs/</tt></a> (available for 32-bit and 64-bit systems). Please note: due to a different Python version, the Debian and Ubuntu .deb packages are different! Please use the correct one for your distribution.
These packages have been tested on up-to-date versions of Karmic Koala (9.10), Lucid Lynx (10.04) and Debian Lenny (5.05). 
Open a terminal and install GimmeMotifs as follows. Go to the directory where you downloaded GimmeMotifs, for example:

<pre>
cd&nbsp;~/Downloads

</pre>
Start the install (substitute the package name for the .deb package you downloaded):

<pre>
sudo&nbsp;dpkg&nbsp;-i&nbsp;gimmemotifs_0.61-1_i386.deb&nbsp;

</pre>
Likely, dpkg will complain about some missing dependencies. Install all dependencies with:

<pre>
sudo&nbsp;apt-get&nbsp;-f&nbsp;install

</pre>
Complete the GimmeMotifs installation with:

<pre>
sudo&nbsp;dpkg&nbsp;-i&nbsp;gimmemotifs_0.61-1_i386.deb&nbsp;

</pre>
Currently, there is a bug with the versions of the Parallel Python (python-pp) and Numpy (python-numpy) in the Ubuntu and Debian repositories. Therefore the package python-pp is not installed as a dependency. This can be fixed by installing the latest version of Parallel Python from the Python Package Index:

<pre>
sudo&nbsp;easy_install&nbsp;pp&nbsp;

</pre>
Now you should have a working version of GimmeMotifs! The next steps are to install additional motif tools (optional, see section <a href="#sec:adding_tools">3.3.5</a>) and to do some configuration (required, see section <a href="#sec:configuration">3.3</a> ). You can also directly try the quick example (section <a href="#sec:quick_example">4.1</a>), if you're impatient (but don't forget to perform the additional steps!)

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.1.2">
3.1.2</a>&nbsp;&nbsp;Installation on Fedora</h4>
Download the latest GimmeMotifs .rpm package from <a href="http://www.ncmls.nl/bioinfo/gimmemotifs/"><tt>http://www.ncmls.nl/bioinfo/gimmemotifs/</tt></a> (available for 32-bit and 64-bit systems). This package has been tested on an up-to-date version of Fedora 13. Install GimmeMotifs as follows (substitute the package name for the .rpm package you downloaded):

<pre>
sudo&nbsp;yum&nbsp;install&nbsp;--nogpgcheck&nbsp;gimmemotifs_0.61-1_i386.rpm&nbsp;

</pre>
GimmeMotifs doesn't play nice with SELinux enabled on Fedora, sorry. To turn it off:

<pre>
sudo&nbsp;setenforce&nbsp;0

</pre>
Now you should have a working version of GimmeMotifs. The next steps are to install additional motif tools (optional, see section <a href="#sec:adding_tools">3.3.5</a>) and to do some configuration (required, see section <a href="#sec:configuration">3.3</a> ). You can also directly try the quick example (section <a href="#sec:quick_example">4.1</a>), if you're impatient (but don't forget to perform the additional steps!)

<div class="p"><!----></div>
     <h3><a name="tth_sEc3.2">
3.2</a>&nbsp;&nbsp;Installation from source</h3>

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.2.1">
3.2.1</a>&nbsp;&nbsp;Prerequisites</h4>
Before you can install GimmeMotifs you'll need: 

<ul>
<li> some Python modules and other packages
<div class="p"><!----></div>
</li>

<li> motif prediction tools
<div class="p"><!----></div>
</li>
</ul>

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.2.2">
3.2.2</a>&nbsp;&nbsp;Required packages (Python)</h4>

<ul>
<li> Python 2.5 or 2.6 (not Python 3) <a href="http://www.python.org"><tt>http://www.python.org</tt></a>
<div class="p"><!----></div>
</li>

<li> Scipy <a href="http://www.scipy.org/"><tt>http://www.scipy.org/</tt></a> <br />
SciPy is the fundamental package needed for scientific computing with Python.
<div class="p"><!----></div>
</li>

<li> matplotlib (0.98 or higher) <a href="http://matplotlib.sourceforge.net/"><tt>http://matplotlib.sourceforge.net/</tt></a> <br />
A python 2D plotting library. All figures and plots produced by GimmeMotifs are made using matplotlib.
<div class="p"><!----></div>
</li>

<li> parallel python 1.6.0 <a href="http://www.parallelpython.com/"><tt>http://www.parallelpython.com/</tt></a> <br />
A python module which provides mechanism for parallel execution of python code. This Python library is used for parallel execution of for instance the motif finding tools.
<div class="p"><!----></div>
</li>

<li> kid <a href="http://www.kid-templating.org/"><tt>http://www.kid-templating.org/</tt></a> <br />
A simple template language for XML based vocabularies; used to produce the HTML reports.
<div class="p"><!----></div>
</li>
</ul>

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.2.3">
3.2.3</a>&nbsp;&nbsp;Other required packages</h4>

<ul>
<li> gsl <a href="http://www.gnu.org/software/gsl/"><tt>http://www.gnu.org/software/gsl/</tt></a> <br />
The GNU Scientific Library. 
This library might already be installed on your system, but you'll also need the development headers to compile GimmeMotifs!.
<div class="p"><!----></div>
</li>
</ul>

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.2.4">
3.2.4</a>&nbsp;&nbsp;Additional motif prediction programs</h4>
A lot of motif prediction tools are compiled and/or installed with GimmeMotifs. The following tools have to be installed seperately: 

<ul>
<li> Weeder <a href="#pavesi_weeder_2004" name="CITEpavesi_weeder_2004">[10</a>] <a href="http://159.149.109.9/modtools/"><tt>http://159.149.109.9/modtools/</tt></a>
<div class="p"><!----></div>
</li>
</ul>
Please consult the respective manuals regarding installation of these tools. It's always possible to install these programs after installation of GimmeMotifs and update the configuration files to include the new tools (see section <a href="#sec:adding_tools">3.3.5</a>). However, during installation, GimmeMotifs will try to find any installed tools and add them automatically, so that's the easiest option.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.2.5">
3.2.5</a>&nbsp;&nbsp;Building from source</h4>
You can download the lastest version of GimmeMotifs at:<br />
<a href="http://www.ncmls.eu/bioinfo/gimmemotifs/"><tt>http://www.ncmls.eu/bioinfo/gimmemotifs/</tt></a>. <br />
<br />
Start by unpacking the source archive

<pre>
tar&nbsp;xvzf&nbsp;gimmemotifs-1.00.tar.gz
cd&nbsp;gimmemotifs-1.00

</pre>
You can build GimmeMotifs with the following command:

<pre>
python&nbsp;setup.py&nbsp;build

</pre>
Run the tests to check if the basics work correctly:

<pre>
python&nbsp;run_tests.py

</pre>
If you encounter no errors, go ahead with installing GimmeMotifs (root privileges required):

<pre>
sudo&nbsp;python&nbsp;setup.py&nbsp;install

</pre>
During installation GimmeMotifs will try to locate the tools you have installed. If you have recently installed them, running an <tt>updatedb</tt> will be necessary. Using this option GimmeMotifs will create a configuration file, the default is:

<pre>
/usr/share/gimmemotifs/gimmemotifs.cfg

</pre>
This is a system-wide configuration that can be used by all users.

<div class="p"><!----></div>
It is also possible to run the <tt>setup.py install</tt> command with the <tt>-prefix</tt>, <tt>-home</tt>, or <tt>-install-data</tt> options, to install in GimmeMotifs in a different location (for instance, in your own home directory). This should be fine, however, these alternative methods of installing GimmeMotifs have not been extensively tested. Please note that in this case the configuration file will be created, but every user will have to put this configuration file in his/her home directory: <tt>/.gimmemotifs.cfg</tt>. The install script will also inform you of this during install. Please contact me if you run into problems with the installation.
Once the installation is finished, you can try the quick example (section <a href="#sec:quick_example">4.1</a>), or continue with the configuration in the next section.

     <h3><a name="tth_sEc3.3">
3.3</a>&nbsp;&nbsp;Configuration</h3>
<a name="sec:configuration">
</a>

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.1">
3.3.1</a>&nbsp;&nbsp;Data sources</h4>
You will need some genome fasta files for any motif-prediction if you want to run GimmeMotifs with BED files as input (which is recommended, although GimmeMotifs will also work on FASTA files). To get from a BED file to the sequence information these genomic fasta files are absolutely required. The fasta files should be organized in one directory with one file per chromosome or scaffold, with the filename being the chromosome name with an extension of <tt>.fa</tt>, <tt>.fsa</tt> or <tt>.fasta</tt>. No exceptions, no different layouts. A good source is the UCSC Genome Browser database <a href="#rhead_ucsc_2010" name="CITErhead_ucsc_2010">[11</a>]. For instance, the human hg18 files needed to run the examples included with GimmeMotifs can be downloaded here:

<pre>
ftp://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/chromFa.zip
&nbsp;
</pre>
All fasta files need to be indexed before GimmeMotifs can use them, see section <a href="#sec:indexing">3.3.2</a>.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.2">
3.3.2</a>&nbsp;&nbsp;Indexing the genomes</h4>
<a name="sec:indexing">
</a>
All the genomes that you want to use with GimmeMotifs will need to be indexed for (relatively) fast retrieval of sequences. You can do this, once you have installed GimmeMotifs, by running the following command (as root or with sudo):

<pre>
create_genome_index.py&nbsp;-f&nbsp;/dir/to/fasta/files/&nbsp;-n&nbsp;genome_name

</pre>
For instance, if I wanted to index the human genome (version hg18) on my computer, where all fasta files are located in the directory <tt>/usr/share/genome/</tt> I would run the following command:

<pre>
sudo&nbsp;create_genome_index.py&nbsp;-f&nbsp;/usr/share/genome/hg18/&nbsp;-n&nbsp;hg18

</pre>
Repeat this step for every additional genome or organism that you want to use GimmeMotifs with. Please note: for Weeder, currently only hg18, hg19, mm9, sacCer2 and xenTro2 are supported as organism names (following the UCSC naming convention). This will be fixed as a configuration file in a later release.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.3">
3.3.3</a>&nbsp;&nbsp;Adding gene files</h4>
When using the <tt>genomic_matched</tt> background setting (which is the default), there needs to be a file describing genes in BED format in the <tt>gene_dir</tt>, which is defined in the configuration file. By default this is: <tt>/usr/share/gimmemotifs/genes/</tt>. The file needs to be named <tt>&lt;index_name&#62;.bed</tt>, so for instance <tt>hg18.bed</tt>. By default <tt>hg18.bed</tt>, <tt>mm9.bed</tt> and <tt>xenTro2.bed</tt> are included.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.4">
3.3.4</a>&nbsp;&nbsp;The easy way: <tt>add_organism.py</tt></h4>
The script <tt>add_organism.py</tt> combines the previous two steps (indexing the fasta files, and adding a gene file), and makes sure the gene BED file is in the correct place with the correct name. This is the easiest way to add a new genome/organism for use with GimmeMotifs.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.5">
3.3.5</a>&nbsp;&nbsp;Adding motif prediction tools</h4> <a name="sec:adding_tools">
</a>
Please note that these steps are only necessary when you have installed any of these tools after you have installed GimmeMotifs.

<div class="p"><!----></div>

<h4>Weeder</h4>
After installing Weeder the following section needs to be added to the GimmeMotifs configuration file:

<pre>
[Weeder]
bin&nbsp;=&nbsp;/usr/share/Weeder/weederTFBS.out
dir&nbsp;=&nbsp;/usr/share/Weeder/&nbsp;

</pre>
All other Weeder binaries should be present in the same directory as <tt>weederTFBS.out</tt>. The directory specified by <tt>dir</tt> should contain the FreqFiles directory included with Weeder.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.6">
3.3.6</a>&nbsp;&nbsp;MotifSampler configuration</h4>
<a name="sec:MotifSampler">
</a>
If you want to use MotifSampler there is one more step that you'll have to take <em>after</em> installation of GimmeMotifs. For every organism, you'll need a MotifSampler background. These can be created with <tt>CreateBackgroundModel</tt> (which is included with GimmeMotifs or can be downloaded from the same site as MotifSampler). The background model file needs to be saved in the directory <tt>/usr/share/gimmemotifs/MotifSampler</tt> and it should be named <tt>&lt;organism_index_name&#62;.bg</tt>. So, for instance, if I downloaded the human epd background (<tt>epd_homo_sapiens_499_chromgenes_non_split_3.bg</tt>), this file should be saved as <tt>/usr/share/gimmemotifs/MotifSampler/hg18.bg</tt>.

<div class="p"><!----></div>
      <h4><a name="tth_sEc3.3.7">
3.3.7</a>&nbsp;&nbsp;Other configuration options</h4>
All of GimmeMotifs configuration is stored in <tt>/usr/share/gimmemotifs/gimmemotifs.cfg</tt> or <tt>/.gimmemotifs.cfg</tt>. If the file <tt>/.gimmemotifs.cfg</tt> exists in your home directory this will always have precedence over the system-wide configuration.
The configuraton file is created at installation time with all defaults set, but you can always edit it afterwards. It contains two sections <tt>main</tt> and <tt>params</tt> that take care of paths, file locations, parameter settings etc. Additionally, every motif tool has it's own section.
Let's have a look at the options.

<pre>
[main]
index_dir&nbsp;=&nbsp;/usr/share/gimmemotifs/genome_index
template_dir&nbsp;=&nbsp;/usr/share/gimmemotifs/templates
seqlogo&nbsp;=&nbsp;/usr/local/bin/seqlogo
score_dir&nbsp;=&nbsp;/usr/share/gimmemotifs/score_dists
motif_databases&nbsp;=&nbsp;/usr/share/gimmemotifs/motif_databases
gene_dir&nbsp;=&nbsp;/usr/share/gimmemotifs/genes
tools_dir&nbsp;=&nbsp;/usr/share/gimmemotifs/tools

</pre>

<ul>
<li> <tt>index_dir</tt> The location of the indeces of the genome fasta-files.
<div class="p"><!----></div>
</li>

<li> <tt>template_dir</tt> The location of the KID html templates, used to generate the reports.
<div class="p"><!----></div>
</li>

<li> <tt>seqlogo</tt> The seqlogo executable.
<div class="p"><!----></div>
</li>

<li> <tt>score_dir</tt> To generate p-values, a pre-calculated file with mean and sd of score distributions is needed. These are located here.
<div class="p"><!----></div>
</li>

<li> <tt>motif_databases</tt> For now contains only the JASPAR motifs.
<div class="p"><!----></div>
</li>

<li> <tt>gene_dir</tt> Directory with bed-files containing gene locations for every indexed organism. This is needed to create the matched genomic background.
<div class="p"><!----></div>
</li>

<li> <tt>tools_dir</tt> Here all tools included with GimmeMotifs are stored.
<div class="p"><!----></div>
</li>
</ul>

<div class="p"><!----></div>

<pre>
[params]
background&nbsp;=&nbsp;genomic_matched,random
use_strand&nbsp;=&nbsp;False
tools&nbsp;=&nbsp;MDmodule,Weeder,MotifSampler
analysis&nbsp;=&nbsp;medium
pvalue&nbsp;=&nbsp;0.001
width&nbsp;=&nbsp;200
fraction&nbsp;=&nbsp;0.2
genome&nbsp;=&nbsp;hg18
lwidth&nbsp;=&nbsp;500
cluster_threshold&nbsp;=&nbsp;0.95
available_tools&nbsp;=&nbsp;Weeder,MDmodule,MotifSampler,gadem,meme,trawler
abs_max&nbsp;=&nbsp;1000
enrichment&nbsp;=&nbsp;1.5
max_time&nbsp;=&nbsp;None

</pre>
This section specifies all the default GimmeMotifs parameters. Most of these can also be specified at the command-line when running GimmeMotifs, in which case they will override the parameters specified here.

<div class="p"><!----></div>
 <h2><a name="tth_sEc4">
4</a>&nbsp;&nbsp;Usage</h2>

     <h3><a name="tth_sEc4.1">
4.1</a>&nbsp;&nbsp;Quick GimmeMotifs example</h3>
<a name="sec:quick_example">
</a>
You can try GimmeMotifs with a small example dataset included in the examples directory, included with GimmeMotifs. This example does not require any additional configuration if GimmeMotifs is installed correctly.

<div class="p"><!----></div>
Change to a directory where you have write permissions and run the following command (substitute the filename with the location of the file on your system):

<pre>
gimme_motifs.py&nbsp;-i&nbsp;/usr/share/gimmemotifs/examples/TAp73alpha.fa&nbsp;-n&nbsp;p73

</pre>

<div class="p"><!----></div>
The <tt>-n</tt> or <tt>-name</tt> option defines the name of the output directory that is created. All output files are stored in this directory.

<div class="p"><!----></div>
Depending on your computer you may have to wait some minutes for your results. Once GimmeMotifs is finished you can open <a href="p73/p73_motif_report.html"><tt>p73/p73_motif_report.html</tt></a> in your browser. 

<div class="p"><!----></div>
     <h3><a name="tth_sEc4.2">
4.2</a>&nbsp;&nbsp;GimmeMotifs example</h3>
This example is the same as above, except it will start from a BED file. This example does require you to have hg18 present and indexed. 
Change to a directory where you have write permissions and run the following command (substitute the filename with the location of the file on your system):

<pre>
gimme_motifs.py&nbsp;-i&nbsp;/usr/share/gimmemotifs/examples/TAp73alpha.bed&nbsp;-n&nbsp;example

</pre>

<div class="p"><!----></div>
The <tt>-n</tt> or <tt>-name</tt> option defines the name of the output directory that is created. All output files are stored in this directory.

<div class="p"><!----></div>
Depending on your computer you may have to wait some minutes for your results. Once GimmeMotifs is finished you can open <a href="example/example_motif_report.html"><tt>example/example_motif_report.html</tt></a> in your browser. 

<div class="p"><!----></div>
     <h3><a name="tth_sEc4.3">
4.3</a>&nbsp;&nbsp;Using GimmeMotifs: best practices and tips</h3>

      <h4><a name="tth_sEc4.3.1">
4.3.1</a>&nbsp;&nbsp;GimmeMotifs is multi-threaded</h4>
GimmeMotifs runs multi-threaded and uses all the CPU's in the system. This means that all the programs will be run in parallel as much as possible. Of course some programs are still single-threaded, and will not benefit from this. Because GimmeMotifs uses all the available CPU's it does not make much sense to start multiple GimmeMotifs jobs at the same time. 

<div class="p"><!----></div>
      <h4><a name="tth_sEc4.3.2">
4.3.2</a>&nbsp;&nbsp;Running time</h4>
The running time of GimmeMotifs greatly depends on which tools you use for prediction and how large the dataset is. Some of the tools might take a very long time and two of them, GADEM and MoAn, are not added to the default tools because of this reason. You can always use them for an analysis (by specifying the <tt>-t</tt> command-line option), but it is recommended to only do this for a small dataset (say, less than 5000 peaks). Weeder in combination with the <tt>xl</tt> analysis can also take a very long time, so be prepared. In general a <tt>small</tt> analysis will be the quickest, and a <tt>xl</tt> analysis will be the slowest.

<div class="p"><!----></div>
While GimmeMotifs is developed specifically for ChIP-seq datasets, most motif prediction tools are not. In practice this means that it does not make much sense to predict motifs on a large amount of sequences, as this will usually not result in higher quality motifs. Therefore GimmeMotifs uses an absolute limit for the prediction set. By default 20% of the sequences are used as input for motif prediction, but with an absolute maximum. This is controlled by the <tt>abs_max</tt> parameter in the configuration file, which is set to 1000 by default. In general, if you have a large amount of peaks, you can also consider to run GimmeMotifs on the top sequences of your input, for instance the 5000 highest peaks. 

<div class="p"><!----></div>
There are two options that you can use to control the running time of GimmeMotifs. First, you can set an absolute time limit with the <tt>max_time</tt> option. This option (in hours) determines the maximum time used for motif prediction. If some programs take longer, the running jobs will be terminated, and the program will continue with all the motifs that have been predicted so far. The other option is kind of an emergency button: when you think that GimmeMotifs has been running long enough, you can press Ctrl+C <b>once, and only once!</b>. This will signal GimmeMotifs to terminate the running jobs and continue with the analysis. Please note that this works almost always, but still, there is a small chance that program might be in a function where the Ctrl-C option screws up, and GimmeMotifs will not be able to handle the result gracefully. 

<div class="p"><!----></div>
      <h4><a name="tth_sEc4.3.3">
4.3.3</a>&nbsp;&nbsp;Intermediate results</h4>
GimmeMotifs produces a lot of intermediate results, such as all predicted motifs, fasta-files used for validation and so on. These are deleted by default (as they can get quite large), but if you are interested in them, you can specify the <tt>-k</tt> option.

<div class="p"><!----></div>
      <h4><a name="tth_sEc4.3.4">
4.3.4</a>&nbsp;&nbsp;Running on FASTA files</h4>
It is also possible to run GimmeMotifs on a FASTA file as input instead of a BED file. This is detected automatically if you're inputfile correctly formatted according to FASTA specifications. In this case it is not possible to generate a genomic matched background, so only the random Markov background will be used. Please note that for best results, all the sequences should be of the same length. This is not necessary for motif prediction, but the statistics and positional preference plots will be wrong if sequences have different lengths. Also see the next section.

<div class="p"><!----></div>
      <h4><a name="tth_sEc4.3.5">
4.3.5</a>&nbsp;&nbsp;Small input sets</h4>
Keep in mind that GimmeMotifs is developed for larger datasets, where you have the luxury to use a large fraction of your input for validation. So, at least several hundred sequences would be optimal. If you want to run GimmeMotifs on a small input dataset, it might be worthwile to increase the fraction used for prediction (with the <tt>-f</tt> parameter.

<div class="p"><!----></div>
     <h3><a name="tth_sEc4.4">
4.4</a>&nbsp;&nbsp;Detailed options</h3>

<ul>
<li> <tt>-i</tt> or <tt>-inputfile</tt>

<div class="p"><!----></div>
This is the only mandatory option. The inputfile needs to be in BED or FASTA format. BED-fomatted files need to contain at least three tab-seperated columns describing chromosome name, start and end. The fourth column is optional, if specified it will be used by MDmodule to sort the features before motif prediction. GimmeMotifs will take the center of these features, and subsequently extend those to the width specified by the <tt>width</tt> parameter (see below).
<div class="p"><!----></div>
</li>

<li> 
<tt>-n</tt> or <tt>-name</tt>

<div class="p"><!----></div>
The name of your analysis. All outputfiles will be stored in a directory named as given by this parameter. By default this will be gimmemotifs_dd_mm_yyyy, where d,m and y are the current day, month and year respectively.
<div class="p"><!----></div>
</li>

<li> 
<tt>-a</tt> or <tt>-analysis</tt>

<div class="p"><!----></div>
The size of motifs to look for: small (5-8), medium (5-12), large (6-15) or xl (6-20). The larger the motifs, the longer GimmeMotifs will run. The 'xl' can take a very long time!
<div class="p"><!----></div>
</li>

<li> 
<tt>-g</tt> or <tt>-genome</tt>

<div class="p"><!----></div>
Name of the genome (index) to use. For instance, for the example in section <a href="#sec:indexing">3.3.2</a> this would be <tt>hg18</tt>.
<div class="p"><!----></div>
</li>

<li> 
<tt>-s</tt> or <tt>-singlestrand</tt>

<div class="p"><!----></div>
Only use the + strand for prediction (off by default).
<div class="p"><!----></div>
</li>

<li> 
<tt>-f</tt> or <tt>-fraction</tt>

<div class="p"><!----></div>
This parameter controls the fraction of the sequences used for prediction. This 0.2 by default, so in this case a randomly chosen 20% of the sequences will be used for prediction. The remaining sequences will be used for validation (enrichment, ROC curves etc.). If you have a large set of sequences (ie. most ChIP-seq peak sets), this is fine. However, if your set is smaller, it might be worthwile to increase this prediction fraction.
<div class="p"><!----></div>
</li>

<li> 
<tt>-w</tt> or <tt>-width</tt>

<div class="p"><!----></div>
This is the width of the sequences used for motif prediction. Smaller sequences will result in a faster analysis, but you are of course limited by the accuracy of your data. For the tested ChIP-seq data sets 200 performs fine.
<div class="p"><!----></div>
</li>

<li> 
<tt>-e</tt> or <tt>-enrichment</tt>

<div class="p"><!----></div>
All motifs should have an absolute enrichment of at least this parameter compared to background to be called significant.
<div class="p"><!----></div>
</li>

<li> 
<tt>-p</tt> or <tt>-pvalue</tt>

<div class="p"><!----></div>
All motifs should have a pvalue of at most this parameter (hypergeometric enrichment compared to background) to be called significant.
<div class="p"><!----></div>
</li>

<li> 
<tt>-b</tt> or <tt>-background</tt>

<div class="p"><!----></div>
Type of background to use. By default <tt>random</tt> (1st order Markov model, similar dinucleotide frequencies as your sequences) and <tt>matched_genomic</tt> (randomly chosen from the genome with a similar distribution respective to the TSS of genes) are used.
<div class="p"><!----></div>
</li>

<li> 
<tt>-l</tt> or <tt>-localization_width</tt>

<div class="p"><!----></div>
Width used in the positional preference plots.
<div class="p"><!----></div>
</li>

<li> 
<tt>-t</tt> or <tt>-tools</tt>

<div class="p"><!----></div>
A comma-seperated list of all the motif prediction tools to use. By default all installed tools that are specified in the GimmeMotifs configuration file are used.
<div class="p"><!----></div>
</li>

<li>
<tt>-max_time</tt>

<div class="p"><!----></div>
 Time limit for motif prediction in hours. Use this to control the maximum number of hours that GimmeMotifs uses for motif prediction. After this time, all jobs that are still running will be terminated, and GimmeMotifs will continue with the motifs that are predicted so far.
<div class="p"><!----></div>
</li>
</ul>

<div class="p"><!----></div>
 <h2><a name="tth_sEc5">
5</a>&nbsp;&nbsp;Other scripts</h2>
In addition to <tt>gimme_motifs.py</tt> the GimmeMotifs package contains several other tools that can perform the various substeps of GimmeMotifs, as well as other useful tools. Run them to see the options.

<div class="p"><!----></div>
     <h3><a name="tth_sEc5.1">
5.1</a>&nbsp;&nbsp;Input formats</h3>
Most tools in this section take a file in PWM format as input. This is actually a file with Position Specific Scoring Matrices (PSSMs) containing <em>frequencies</em>. It looks like this:

<pre>
&#62;motif1
0.3611	0.0769	0.4003	0.1664
0.2716	0.0283	0.5667	0.1381
0.6358	0.0016	0.3344	0.0330
0.0016	0.9859	0.0016	0.0157
0.8085	0.0063	0.0502	0.1397
&#62;motif2
0.2276	0.0157	0.0330	0.7284
0.0031	0.0016	0.9984	0.0016
0.0377	0.3799	0.0016	0.5856
0.0816	0.7096	0.0173	0.1962
0.1350	0.4035	0.0675	0.3987

</pre>
The frequencies are seperated by tabs, and in the order A,C,G,T.

<div class="p"><!----></div>
     <h3><a name="tth_sEc5.2">
5.2</a>&nbsp;&nbsp;Descriptions</h3>

<h3>closest_motif_match.py</h3>
Taking an input file with motifs, find the best matching file in another file of motifs (according to the WIC metric).

<div class="p"><!----></div>

<h3>create_genome_index.py</h3>
Creates an index to use with GimmeMotifs. See section <a href="#sec:configuration">3.3</a> for details.

<div class="p"><!----></div>

<h3>generate_sequences.py</h3>
Generate random sequences with the same dinucleotide distribution as the input sequences according to a 1st order Markov model trained on the input sequences. The <tt>-n</tt> options is set to 10 by default. The length distribution of the sequences in the outputfile will be similar as the inputfile.

<div class="p"><!----></div>

<h3>motif_cluster.py</h3>
Cluster a set of motifs with the WIC metric.

<div class="p"><!----></div>

<h3>motif_localization_plots.py</h3>
Create the positional preference plots for all the motifs in the input PWM file. This will give best results if all the sequences in the FASTA-formatted inputfile have the same length. Keep in mind that this only makes sense if the sequences are centered around a similar feature (transcription start site, highest point in a peak, etc.). The default threshold for motif scanning is 0.95, see <tt>pwmscan.py</tt> for more details.

<div class="p"><!----></div>

<h3>motif_roc.py</h3>
Given a sample (positives, peaks) and a background file (random sequences, random promoters or similar),  creates a ROC plot for all the motifs in an input PWM file. All the motifs will be plotted in the same graph, you can select one or more specific motifs to plot with the <tt>-i</tt> option.

<div class="p"><!----></div>

<h3>motif_roc_metrics.py</h3>
Given a sample (positives, peaks) and a background file (random sequences, random promoters or similar),  calculated several statistics for all the motifs in an input PWM file. You can select one or more specific motifs with the <tt>-i</tt> option. The ROC area under curve (ROC_AUC), Mean Normalized Conditional Probability (MNCP) and maximum f-measure are reported for each motif.

<div class="p"><!----></div>

<h3>pwm2logo.py</h3>
Convert the motifs in a PWM file to a logo using weblogo.

<div class="p"><!----></div>

<h3>pwmscan.py</h3>
Scan a set of sequences with a set of motifs, and give the resulting matches in GFF or BED format. The threshold is based on the maximum and minimum possible score for each motif. So, 0.95 means that the score of a motif should be at least 95% of the (maximum score - minimum score). This should probably not be set much lower than 0.8, and should be generally at least 0.9 for good specificity. Keep in mind that the optimal threshold might be different for each motif!

<div class="p"><!----></div>

<h3>track2fasta.py </h3>
Convert a set of BED formatted sequences to a FASTA file. The genome needs to be indexed for GimmeMotifs using <tt>create_genome_index.py</tt>.

<div class="p"><!----></div>

<div class="p"><!----></div>
 <h2><a name="tth_sEc6">
6</a>&nbsp;&nbsp;Acknowledgements</h2>
We are grateful to Waseem Akhtar, Robert Akkers, Max Koeppel, Evelyn Kouwenhoven, Marion Lohrum, Leonie Smeenk and Jo Zhou for providing data and feedback during GimmeMotifs development. Also we would like to thank Stefanie Bartels, Adalberto Costessi, Joost Martens and Nagesha Rao for testing and helpful discussion.
Of course GimmeMotifs by itself wouldn't be able to do anything, if there wasn't such a number of excellent tools available. Therefore, a big thanks to all the authors of the motif prediction programs for making their software publicly available and allowing me to distribute them with GimmeMotifs!

<div class="p"><!----></div>

<h2>References</h2>

<dl compact="compact">
 <dt><a href="#CITEliu_bioprospector:_2001" name="liu_bioprospector:_2001">[1]</a></dt><dd>
X&nbsp;Liu, D&nbsp;L Brutlag, and J&nbsp;S Liu.
 BioProspector: discovering conserved DNA motifs in upstream
  regulatory regions of co-expressed genes.
 <em>Pacific Symposium on Biocomputing. Pacific Symposium on
  Biocomputing</em>, pages 127-138, 2001.
 PMID: 11262934.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEli_gadem:genetic_2009" name="li_gadem:genetic_2009">[2]</a></dt><dd>
Leping Li.
 GADEM: a genetic algorithm guided formation of spaced dyads coupled
  with an EM algorithm for motif discovery.
 <em>Journal of computational biology : a journal of computational
  molecular cell biology</em>, 16 (2): 317-329, February 2009.
 PMID: 19193149 PMCID: 2756050.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEao_environmentally_2004" name="ao_environmentally_2004">[3]</a></dt><dd>
Wanyuan Ao, Jeb Gaudet, W.&nbsp;James Kent, Srikanth Muttumu, and Susan&nbsp;E. Mango.
 Environmentally induced foregut remodeling by PHA-4/FoxA and
  DAF-12/NHR.
 <em>Science</em>, 305 (5691): 1743-1746, September
  2004.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEliu_algorithm_2002" name="liu_algorithm_2002">[4]</a></dt><dd>
X.&nbsp;Shirley Liu, Douglas&nbsp;L. Brutlag, and Jun&nbsp;S. Liu.
 An algorithm for finding protein-DNA binding sites with
  applications to chromatin-immunoprecipitation microarray experiments.
 <em>Nat Biotech</em>, 20 (8): 835-839, 2002.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEbailey_meme_2009" name="bailey_meme_2009">[5]</a></dt><dd>
Timothy&nbsp;L. Bailey, Mikael Boden, Fabian&nbsp;A. Buske, Martin Frith, Charles&nbsp;E.
  Grant, Luca Clementi, Jingyuan Ren, Wilfred&nbsp;W. Li, and William&nbsp;S. Noble.
 MEME SUITE: tools for motif discovery and searching.
 <em>Nucl. Acids Res.</em>, 37 (suppl_2): W202-208,
  July 2009.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEvalen_discovery_2009" name="valen_discovery_2009">[6]</a></dt><dd>
Eivind Valen, Albin Sandelin, Ole Winther, and Anders Krogh.
 Discovery of regulatory elements is improved by a discriminatory
  approach.
 <em>PLoS Comput Biol</em>, 5 (11): e1000562,
  November 2009.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEthijs_higher-order_2001" name="thijs_higher-order_2001">[7]</a></dt><dd>
G&nbsp;Thijs, M&nbsp;Lescot, K&nbsp;Marchal, S&nbsp;Rombauts, B&nbsp;De Moor, P&nbsp;Rouzé, and Y&nbsp;Moreau.
 A higher-order background model improves the detection of promoter
  regulatory elements by gibbs sampling.
 <em>Bioinformatics (Oxford, England)</em>, 17 (12):
  1113-1122, December 2001.
 PMID: 11751219.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEthijs_gibbs_2002" name="thijs_gibbs_2002">[8]</a></dt><dd>
Gert Thijs, Kathleen Marchal, Magali Lescot, Stephane Rombauts, Bart&nbsp;De Moor,
  Pierre Rouzé, and Yves Moreau.
 A gibbs sampling method to detect overrepresented motifs in the
  upstream regions of coexpressed genes.
 <em>Journal of Computational Biology: A Journal of Computational
  Molecular Cell Biology</em>, 9 (2): 447-464, 2002.
 PMID: 12015892.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEettwiller_trawler:_2007" name="ettwiller_trawler:_2007">[9]</a></dt><dd>
Laurence Ettwiller, Benedict Paten, Mirana Ramialison, Ewan Birney, and Joachim
  Wittbrodt.
 Trawler: de novo regulatory motif discovery pipeline for chromatin
  immunoprecipitation.
 <em>Nat Meth</em>, 4 (7): 563-565, July 2007.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITEpavesi_weeder_2004" name="pavesi_weeder_2004">[10]</a></dt><dd>
Giulio Pavesi, Paolo Mereghetti, Giancarlo Mauri, and Graziano Pesole.
 Weeder web: discovery of transcription factor binding sites in a set
  of sequences from co-regulated genes.
 <em>Nucl. Acids Res.</em>, 32 (suppl_2): W199-203,
  July 2004.

<div class="p"><!----></div>
</dd>
 <dt><a href="#CITErhead_ucsc_2010" name="rhead_ucsc_2010">[11]</a></dt><dd>
Brooke Rhead, Donna Karolchik, Robert&nbsp;M. Kuhn, Angie&nbsp;S. Hinrichs, Ann&nbsp;S. Zweig,
  Pauline&nbsp;A. Fujita, Mark Diekhans, Kayla&nbsp;E. Smith, Kate&nbsp;R. Rosenbloom,
  Brian&nbsp;J. Raney, Andy Pohl, Michael Pheasant, Laurence&nbsp;R. Meyer, Katrina
  Learned, Fan Hsu, Jennifer Hillman-Jackson, Rachel&nbsp;A. Harte, Belinda
  Giardine, Timothy&nbsp;R. Dreszer, Hiram Clawson, Galt&nbsp;P. Barber, David Haussler,
  and W.&nbsp;James Kent.
 The UCSC genome browser database: update 2010.
 <em>Nucl. Acids Res.</em>, 38 (suppl_1): D613-619,
  January 2010.</dd>
</dl>


<div class="p"><!----></div>

<br /><br /><hr /><small>File translated from
T<sub><font size="-1">E</font></sub>X
by <a href="http://hutchinson.belmont.ma.us/tth/">
T<sub><font size="-1">T</font></sub>H</a>,
version 3.77.<br />On  1 Oct 2010, 14:12.</small>
</html>
